runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/net/url/QueryParameters.kt

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.net.url
+
+import aws.smithy.kotlin.runtime.collections.MultiMap
+import aws.smithy.kotlin.runtime.collections.MutableMultiMap
+import aws.smithy.kotlin.runtime.collections.mutableMultiMapOf
+import aws.smithy.kotlin.runtime.collections.views.asView
+import aws.smithy.kotlin.runtime.text.encoding.Encodable
+import aws.smithy.kotlin.runtime.text.encoding.PercentEncoding
+
+/**
+ * Represents the parameters in a URL query string.
+ */
+public class QueryParameters private constructor(
```

Comment:
**correctness**: This representation no longer allows for multiple values to be bound to the same parameter name

e.g. `?foo=1&foo=2&foo=3`

That was one of the reasons for sharing `ValuesMap` between headers and query is both have similar semantics between single values, multiple values and whether to set/override vs append. 

Unless I'm missing something (possible) I *think* this representation may be flawed currently.

Reply:
I *think* it's okay, because the query string keys must be case-sensitively unique according to Smithy. https://smithy.io/2.0/spec/http-bindings.html#httpquery-trait

Reply:
Right but the key is unique, this is for multiple _values_ assigned to a key.

Reply:
This representation does support multiple values bound to the same parameter name and there are multiple unit/integ tests which verify this (e.g., [**DefaultCanonicalizerTest**](https://github.com/awslabs/smithy-kotlin/pull/989/files#diff-510c6ad126e119e323ae8e8d5e25622fec6f0e975a4e17288049cbd27e1f407f)'s `testCanonicalQueryParams`).

The `QueryParameters` class implements `MultiMap<K, V>`, which itself extends `Map<K, List<V>>`. There are various convenience members which support things like putting/adding a single value to a key but under the hood values are stored in lists and support the full range of list-based operations at the map level.

Reply:
Ahh ok I missed that part.